File: shared/tools/registry.py
"""
Unified Tool Registry for cross-language tool management
Supports both DGMO and DGM tools with seamless integration
"""

import asyncio
from typing import Dict, List, Optional, Any, Callable, Union
from pathlib import Path
import importlib.util
import sys
import json
from datetime import datetime

from ..types.python.tool import (
    Tool,
    ToolCategory,
    ToolFilter,
    ToolHandler,
    Language,
    ToolContext,
    ToolError
)
from ..types.python.base import Result, ErrorInfo, Metadata
from .python_adapter import PythonTypeScriptAdapter, TypeScriptToolInfo, TypeScriptToolRegistration
from .type_converter import TypeConverter


# DGM Tool Interface (from Python)
class DGMTool:
    """DGM tool definition"""
    def __init__(self, name: str, description: str, input_schema: Dict[str, Any], 
                 output_schema: Optional[Dict[str, Any]] = None, 
                 category: Optional[str] = None, tags: Optional[List[str]] = None):
        self.name = name
        self.description = description
        self.input_schema = input_schema
        self.output_schema = output_schema
        self.category = category
        self.tags = tags


# DGMO Tool Interface (from TypeScript)
class DGMOTool:
    """DGMO tool definition"""
    def __init__(self, id: str, description: str, parameters: Any, execute: Callable):
        self.id = id
        self.description = description
        self.parameters = parameters
        self.execute = execute


class ToolRegistration:
    """Tool registration information"""
    def __init__(self, tool: Tool, handler: ToolHandler, source: str = 'local', 
                 module: Optional[str] = None, original_tool: Optional[Union[DGMOTool, DGMTool]] = None):
        self.tool = tool
        self.handler = handler
        self.source = source
        self.module = module
        self.original_tool = original_tool


class UnifiedToolRegistry:
    """Unified registry for tools across languages"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if not hasattr(self, '_initialized') or not self._initialized:
            self.tools: Dict[str, Dict[Language, ToolRegistration]] = {}
            self._initialized = False
    
    async def initialize(self) -> None:
        """Initialize the registry"""
        if self._initialized:
            return
        
        # Initialize adapters
        await PythonTypeScriptAdapter.initialize()
        
        # Load built-in tools
        await self._load_built_in_tools()
        
        self._initialized = True
    
    async def register(self, tool: Tool, handler: ToolHandler, module: Optional[str] = None, 
                      original_tool: Optional[Union[DGMOTool, DGMTool]] = None) -> None:
        """Register a tool with unified interface"""
        language_map = self.tools.get(tool.id, {})
        
        source = 'local'
        if original_tool:
            source = 'dgmo' if isinstance(original_tool, DGMOTool) else 'dgm'
        elif module:
            source = 'remote'
        
        language_map[tool.language] = ToolRegistration(
            tool=tool,
            handler=handler,
            source=source,
            module=module,
            original_tool=original_tool
        )
        
        self.tools[tool.id] = language_map
        
        # If it's a TypeScript tool, make it available to Python
        if tool.language == Language.TYPESCRIPT and module:
            await PythonTypeScriptAdapter.register_typescript_tool(
                TypeScriptToolRegistration(
                    module=module,
                    tool_id=tool.id,
                    info=TypeScriptToolInfo(
                        id=tool.id,
                        description=tool.description,
                        parameters=tool.input_schema
                    )
                )
            )
    
    async def register_dgm_tool(self, tool_def: DGMTool, module: str) -> None:
        """Register a DGM tool (from Python)"""
        tool = Tool(
            id=tool_def.name,
            name=tool_def.name,
            description=tool_def.description,
            version='1.0.0',
            category=ToolCategory(tool_def.category) if tool_def.category else self._infer_category(tool_def.name),
            language=Language.PYTHON,
            input_schema=tool_def.input_schema,
            output_schema=tool_def.output_schema,
            metadata={
                'id': f'dgm-{tool_def.name}',
                'version': '1.0.0',
                'timestamp': datetime.now().isoformat(),
                'source': 'dgm',
                'tags': tool_def.tags
            }
        )
        
        async def handler(input_data: Any, context: ToolContext) -> Result:
            try:
                # Load and execute the Python tool
                spec = importlib.util.spec_from_file_location("tool_module", module)
                if not spec or not spec.loader:
                    raise Exception(f"Failed to load module {module}")
                
                tool_module = importlib.util.module_from_spec(spec)
                spec.loader.exec_module(tool_module)
                
                # Execute the tool
                if hasattr(tool_module, 'tool_function_async'):
                    result = await tool_module.tool_function_async(input_data, context)
                elif hasattr(tool_module, 'tool_function'):
                    result = await asyncio.get_event_loop().run_in_executor(
                        None, tool_module.tool_function, input_data
                    )
                else:
                    raise Exception(f"No tool function found in {module}")
                
                return Result(
                    success=True,
                    data=result,
                    metadata={
                        'id': context.message_id,
                        'version': '1.0.0',
                        'timestamp': datetime.now().isoformat(),
                        'source': 'dgm'
                    }
                )
            except Exception as e:
                return Result(
                    success=False,
                    error=ErrorInfo(
                        code='DGM_TOOL_ERROR',
                        message=str(e),
                        recoverable=True,
                        details={'tool': tool_def.name, 'error': str(e)}
                    )
                )
        
        await self.register(tool, handler, module, tool_def)
    
    async def register_dgmo_tool(self, tool_def: DGMOTool) -> None:
        """Register a DGMO tool (from TypeScript)"""
        tool = Tool(
            id=tool_def.id,
            name=tool_def.id,
            description=tool_def.description,
            version='1.0.0',
            category=self._infer_category(tool_def.id),
            language=Language.TYPESCRIPT,
            input_schema=self._convert_dgmo_parameters_to_json_schema(tool_def.parameters),
            metadata={
                'id': f'dgmo-{tool_def.id}',
                'version': '1.0.0',
                'timestamp': datetime.now().isoformat(),
                'source': 'dgmo'
            }
        )
        
        async def handler(input_data: Any, context: ToolContext) -> Result:
            try:
                # Convert Python input to TypeScript format
                ts_input = TypeConverter.python_to_typescript(input_data)
                
                # Execute via TypeScript adapter
                result = await PythonTypeScriptAdapter.execute_typescript_tool(
                    tool_def.id,
                    ts_input,
                    {
                        'sessionId': context.session_id,
                        'messageId': context.message_id,
                        'timeout': context.timeout
                    }
                )
                
                # Convert result back to Python
                py_result = TypeConverter.typescript_to_python(result)
                
                return Result(
                    success=True,
                    data=py_result,
                    metadata={
                        'id': context.message_id,
                        'version': '1.0.0',
                        'timestamp': datetime.now().isoformat(),
                        'source': 'dgmo'
                    }
                )
            except Exception as e:
                return Result(
                    success=False,
                    error=ErrorInfo(
                        code='DGMO_TOOL_ERROR',
                        message=str(e),
                        recoverable=True,
                        details={'tool': tool_def.id, 'error': str(e)}
                    )
                )
        
        await self.register(tool, handler, None, tool_def)
    
    async def unregister(self, tool_id: str, language: Optional[Language] = None) -> None:
        """Unregister a tool"""
        if language:
            language_map = self.tools.get(tool_id)
            if language_map and language in language_map:
                del language_map[language]
                if not language_map:
                    del self.tools[tool_id]
        else:
            if tool_id in self.tools:
                del self.tools[tool_id]
    
    async def get(self, tool_id: str, language: Optional[Language] = None) -> Optional[Tool]:
        """Get a tool by ID and optionally language"""
        language_map = self.tools.get(tool_id)
        if not language_map:
            return None
        
        if language:
            registration = language_map.get(language)
            return registration.tool if registration else None
        
        # Return the first available tool
        first_registration = next(iter(language_map.values()), None)
        return first_registration.tool if first_registration else None
    
    def get_handler(self, tool_id: str, language: Language) -> Optional[ToolHandler]:
        """Get tool handler"""
        language_map = self.tools.get(tool_id)
        if not language_map:
            return None
        
        registration = language_map.get(language)
        return registration.handler if registration else None
    
    async def list(self, filter: Optional[ToolFilter] = None) -> List[Tool]:
        """List tools with optional filter"""
        tools = []
        
        for language_map in self.tools.values():
            for registration in language_map.values():
                tool = registration.tool
                
                # Apply filters
                if filter:
                    if filter.category and tool.category != filter.category:
                        continue
                    if filter.language and tool.language != filter.language:
                        continue
                    if filter.tags and filter.tags:
                        tool_tags = tool.metadata.get('tags', []) if tool.metadata else []
                        if not any(tag in tool_tags for tag in filter.tags):
                            continue
                
                tools.append(tool)
        
        return tools
    
    async def search(self, query: str) -> List[Tool]:
        """Search tools by query"""
        lower_query = query.lower()
        tools = []
        
        for language_map in self.tools.values():
            for registration in language_map.values():
                tool = registration.tool
                
                # Search in name, description, and category
                if (lower_query in tool.name.lower() or
                    lower_query in tool.description.lower() or
                    lower_query in tool.category.value.lower()):
                    tools.append(tool)
        
        return tools
    
    async def execute(self, tool_id: str, input_data: Any, context: ToolContext, 
                     preferred_language: Optional[Language] = None) -> Result:
        """Execute a tool with automatic language selection"""
        language_map = self.tools.get(tool_id)
        if not language_map:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='TOOL_NOT_FOUND',
                    message=f'Tool {tool_id} not found',
                    recoverable=False
                )
            )
        
        # Try preferred language first
        if preferred_language:
            registration = language_map.get(preferred_language)
            if registration:
                return await registration.handler(input_data, context)
        
        # Fall back to first available
        first_registration = next(iter(language_map.values()), None)
        if first_registration:
            return await first_registration.handler(input_data, context)
        
        return Result(
            success=False,
            error=ErrorInfo(
                code='NO_HANDLER_FOUND',
                message=f'No handler found for tool {tool_id}',
                recoverable=False
            )
        )
    
    async def validate_input(self, tool_id: str, input_data: Any, 
                           language: Optional[Language] = None) -> Result:
        """Validate tool input against schema"""
        tool = await self.get(tool_id, language)
        if not tool:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='TOOL_NOT_FOUND',
                    message=f'Tool {tool_id} not found',
                    recoverable=False
                )
            )
        
        try:
            # Validate against JSON schema
            valid, error = TypeConverter.validate_against_schema(input_data, tool.input_schema)
            if not valid:
                return Result(
                    success=False,
                    error=ErrorInfo(
                        code='VALIDATION_ERROR',
                        message=error or 'Input validation failed',
                        recoverable=False,
                        details={'input': input_data, 'schema': tool.input_schema}
                    )
                )
            
            return Result(success=True, data=input_data)
        except Exception as e:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='VALIDATION_ERROR',
                    message=str(e),
                    recoverable=False,
                    details={'error': str(e)}
                )
            )
    
    async def discover(self) -> Dict[str, Any]:
        """Get tool discovery information"""
        stats = {
            'total_tools': 0,
            'by_language': {},
            'by_category': {},
            'by_source': {}
        }
        
        for language_map in self.tools.values():
            for language, registration in language_map.items():
                stats['total_tools'] += 1
                
                # Count by language
                lang_str = language.value
                stats['by_language'][lang_str] = stats['by_language'].get(lang_str, 0) + 1
                
                # Count by category
                category_str = registration.tool.category.value
                stats['by_category'][category_str] = stats['by_category'].get(category_str, 0) + 1
                
                # Count by source
                source = registration.source
                stats['by_source'][source] = stats['by_source'].get(source, 0) + 1
        
        return stats
    
    async def _load_built_in_tools(self) -> None:
        """Load built-in tools"""
        # Load Python tools from DGM
        await self._load_dgm_tools()
        
        # Load TypeScript tools from OpenCode
        await self._load_opencode_tools()
    
    async def _load_dgm_tools(self) -> None:
        """Load DGM tools"""
        try:
            tool_modules = [
                {'path': Path(__file__).parent.parent.parent / 'dgm' / 'tools' / 'bash.py', 'name': 'bash'},
                {'path': Path(__file__).parent.parent.parent / 'dgm' / 'tools' / 'edit.py', 'name': 'edit'}
            ]
            
            for module_info in tool_modules:
                module_path = module_info['path']
                if module_path.exists():
                    try:
                        await self._load_python_tool(str(module_path), module_info['name'])
                    except Exception as e:
                        print(f"Failed to load DGM tool from {module_path}: {e}")
        except Exception as e:
            print(f"Failed to load DGM tools: {e}")
    
    async def _load_python_tool(self, module_path: str, tool_name: str) -> None:
        """Load a Python tool module"""
        spec = importlib.util.spec_from_file_location("tool_module", module_path)
        if not spec or not spec.loader:
            return
        
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        
        # Check for tool_info function
        if hasattr(module, 'tool_info'):
            info = module.tool_info()
            
            # Create DGM tool definition
            dgm_tool = DGMTool(
                name=tool_name,
                description=info.get('description', ''),
                input_schema=info.get('input_schema', {}),
                output_schema=info.get('output_schema'),
                category=info.get('category'),
                tags=info.get('tags')
            )
            
            await self.register_dgm_tool(dgm_tool, module_path)
    
    async def _load_opencode_tools(self) -> None:
        """Load TypeScript tools from OpenCode"""
        try:
            tool_modules = [
                '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/opencode/packages/opencode/src/tool/bash.ts',
                '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/opencode/packages/opencode/src/tool/edit.ts',
                '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/opencode/packages/opencode/src/tool/read.ts',
                '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/opencode/packages/opencode/src/tool/write.ts'
            ]
            
            for module_path in tool_modules:
                try:
                    # Load TypeScript tool info via adapter
                    tool_info = await PythonTypeScriptAdapter.load_typescript_module(module_path)
                    if tool_info:
                        # Create DGMO tool definition
                        dgmo_tool = DGMOTool(
                            id=tool_info.get('id', Path(module_path).stem),
                            description=tool_info.get('description', ''),
                            parameters=tool_info.get('parameters', {}),
                            execute=None  # Will be handled by adapter
                        )
                        
                        await self.register_dgmo_tool(dgmo_tool)
                except Exception as e:
                    print(f"Failed to load OpenCode tool from {module_path}: {e}")
        except Exception as e:
            print(f"Failed to load OpenCode tools: {e}")
    
    def _infer_category(self, tool_id: str) -> ToolCategory:
        """Infer tool category from ID"""
        category_map = {
            'bash': ToolCategory.UTILITY,
            'edit': ToolCategory.FILE_SYSTEM,
            'read': ToolCategory.FILE_SYSTEM,
            'write': ToolCategory.FILE_SYSTEM,
            'ls': ToolCategory.FILE_SYSTEM,
            'grep': ToolCategory.TEXT_PROCESSING,
            'glob': ToolCategory.FILE_SYSTEM,
            'patch': ToolCategory.FILE_SYSTEM,
            'multiedit': ToolCategory.FILE_SYSTEM
        }
        
        return category_map.get(tool_id, ToolCategory.UTILITY)
    
    def _convert_dgmo_parameters_to_json_schema(self, parameters: Any) -> Dict[str, Any]:
        """Convert DGMO parameters to JSON Schema"""
        # If it's already a JSON schema, return it
        if isinstance(parameters, dict) and 'type' in parameters:
            return parameters
        
        # Default fallback
        return {
            'type': 'object',
            'properties': {},
            'additionalProperties': True
        }
    
    def get_available_languages(self, tool_id: str) -> List[Language]:
        """Get available languages for a tool"""
        language_map = self.tools.get(tool_id)
        if not language_map:
            return []
        
        return list(language_map.keys())
    
    def supports_language(self, tool_id: str, language: Language) -> bool:
        """Check if a tool supports a specific language"""
        language_map = self.tools.get(tool_id)
        if not language_map:
            return False
        
        return language in language_map


# Create singleton instance
tool_registry = UnifiedToolRegistry()


# Convenience functions
async def register_tool(tool: Tool, handler: ToolHandler, module: Optional[str] = None) -> None:
    """Register a tool in the unified registry"""
    await tool_registry.register(tool, handler, module)


async def get_tool(tool_id: str, language: Optional[Language] = None) -> Optional[Tool]:
    """Get a tool from the registry"""
    return await tool_registry.get(tool_id, language)


async def list_tools(filter: Optional[ToolFilter] = None) -> List[Tool]:
    """List all available tools"""
    return await tool_registry.list(filter)


async def search_tools(query: str) -> List[Tool]:
    """Search for tools"""
    return await tool_registry.search(query)


async def execute_tool(tool_id: str, input_data: Any, context: ToolContext, 
                      preferred_language: Optional[Language] = None) -> Result:
    """Execute a tool"""
    return await tool_registry.execute(tool_id, input_data, context, preferred_language)

File: shared/tools/registry.ts
/**
 * Unified Tool Registry for cross-language tool management
 * Supports both DGMO and DGM tools with seamless integration
 */

import {
  Tool,
  ToolCategory,
  ToolFilter,
  ToolHandler,
  ToolContext,
  ToolError,
} from '../types/typescript/tool.types';
import { Language, Result, ErrorInfo, Metadata } from '../types/typescript/base.types';
import { TypeScriptPythonAdapter } from './typescript-adapter';
import { TypeConverter } from './type-converter';

// DGMO Tool Interface (from OpenCode)
export interface DGMOTool {
  id: string;
  description: string;
  parameters: any; // Zod schema or similar
  execute(
    args: any,
    ctx: DGMOContext,
  ): Promise<{
    metadata: Record<string, any>;
    output: string;
  }>;
}

export interface DGMOContext {
  sessionID: string;
  messageID: string;
  abort: AbortSignal;
  metadata(meta: Record<string, any>): void;
}

// DGM Tool Interface (from Python)
export interface DGMTool {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  outputSchema?: Record<string, unknown>;
  category?: string;
  tags?: string[];
}

// Unified Tool Registration
export interface ToolRegistration {
  tool: Tool;
  handler: ToolHandler;
  source: 'local' | 'remote' | 'dgmo' | 'dgm';
  module?: string;
  originalTool?: DGMOTool | DGMTool;
}

export class UnifiedToolRegistry {
  private static instance: UnifiedToolRegistry;
  private tools = new Map<string, Map<Language, ToolRegistration>>();
  private initialized = false;

  private constructor() {}

  static getInstance(): UnifiedToolRegistry {
    if (!this.instance) {
      this.instance = new UnifiedToolRegistry();
    }
    return this.instance;
  }

  async initialize(): Promise<void> {
    if (this.initialized) {
      return;
    }

    // Initialize adapters
    await TypeScriptPythonAdapter.initialize();

    // Load built-in tools
    await this.loadBuiltInTools();

    this.initialized = true;
  }

  /**
   * Register a tool with unified interface
   */
  async register(
    tool: Tool,
    handler: ToolHandler,
    module?: string,
    originalTool?: DGMOTool | DGMTool,
  ): Promise<void> {
    const languageMap = this.tools.get(tool.id) || new Map<Language, ToolRegistration>();

    languageMap.set(tool.language, {
      tool,
      handler,
      source: originalTool
        ? isDGMOTool(originalTool)
          ? 'dgmo'
          : 'dgm'
        : module
          ? 'remote'
          : 'local',
      module,
      originalTool,
    });

    this.tools.set(tool.id, languageMap);

    // If it's a Python tool, make it available to TypeScript
    if (tool.language === 'python' && module) {
      await TypeScriptPythonAdapter.registerPythonTool({
        module,
        info: {
          name: tool.id,
          description: tool.description,
          input_schema: tool.inputSchema,
        },
      });
    }
  }

  /**
   * Register a DGMO tool (from OpenCode)
   */
  async registerDGMOTool(toolDef: DGMOTool): Promise<void> {
    const tool: Tool = {
      id: toolDef.id,
      name: toolDef.id,
      description: toolDef.description || '',
      version: '1.0.0',
      category: this.inferCategory(toolDef.id),
      language: 'typescript',
      inputSchema: this.convertDGMOParametersToJsonSchema(toolDef.parameters),
      metadata: {
        id: `dgmo-${toolDef.id}`,
        version: '1.0.0',
        timestamp: new Date().toISOString(),
        source: 'dgmo',
      },
    };

    const handler: ToolHandler = async (input: any, context: ToolContext): Promise<Result<any>> => {
      try {
        // Create DGMO context from unified context
        const dgmoContext: DGMOContext = {
          sessionID: context.sessionId,
          messageID: context.messageId,
          abort: context.abortSignal,
          metadata: (meta: Record<string, any>) => {
            Object.entries(meta).forEach(([key, value]) => {
              context.metadata.set(key, value);
            });
          },
        };

        const result = await toolDef.execute(input, dgmoContext);

        return {
          success: true,
          data: result,
          metadata: {
            id: context.messageId,
            version: '1.0.0',
            timestamp: new Date().toISOString(),
            source: 'dgmo',
            ...Object.fromEntries(context.metadata),
          },
        };
      } catch (error) {
        const errorInfo: ErrorInfo = {
          code: 'DGMO_TOOL_ERROR',
          message: error instanceof Error ? error.message : String(error),
          recoverable: true,
          details: error,
        };

        return {
          success: false,
          error: errorInfo,
        };
      }
    };

    await this.register(tool, handler, undefined, toolDef);
  }

  /**
   * Register a DGM tool (from Python)
   */
  async registerDGMTool(toolDef: DGMTool, module: string): Promise<void> {
    const tool: Tool = {
      id: toolDef.name,
      name: toolDef.name,
      description: toolDef.description,
      version: '1.0.0',
      category: (toolDef.category as ToolCategory) || this.inferCategory(toolDef.name),
      language: 'python',
      inputSchema: toolDef.inputSchema,
      outputSchema: toolDef.outputSchema,
      metadata: {
        id: `dgm-${toolDef.name}`,
        version: '1.0.0',
        timestamp: new Date().toISOString(),
        source: 'dgm',
        tags: toolDef.tags,
      },
    };

    const handler: ToolHandler = async (input: any, context: ToolContext): Promise<Result<any>> => {
      try {
        // Convert TypeScript input to Python format
        const pythonInput = TypeConverter.typeScriptToPython(input);

        // Execute via adapter
        const result = await TypeScriptPythonAdapter.executePythonTool(
          module,
          toolDef.name,
          pythonInput,
          {
            sessionId: context.sessionId,
            messageId: context.messageId,
            timeout: context.timeout,
          },
        );

        // Convert Python result back to TypeScript
        const tsResult = TypeConverter.pythonToTypeScript(result);

        return {
          success: true,
          data: tsResult,
          metadata: {
            id: context.messageId,
            version: '1.0.0',
            timestamp: new Date().toISOString(),
            source: 'dgm',
          },
        };
      } catch (error) {
        const errorInfo: ErrorInfo = {
          code: 'DGM_TOOL_ERROR',
          message: error instanceof Error ? error.message : String(error),
          recoverable: true,
          details: error,
        };

        return {
          success: false,
          error: errorInfo,
        };
      }
    };

    await this.register(tool, handler, module, toolDef);
  }

  /**
   * Unregister a tool
   */
  async unregister(toolId: string, language?: Language): Promise<void> {
    if (language) {
      const languageMap = this.tools.get(toolId);
      if (languageMap) {
        languageMap.delete(language);
        if (languageMap.size === 0) {
          this.tools.delete(toolId);
        }
      }
    } else {
      this.tools.delete(toolId);
    }
  }

  /**
   * Get a tool by ID and optionally language
   */
  async get(toolId: string, language?: Language): Promise<Tool | undefined> {
    const languageMap = this.tools.get(toolId);
    if (!languageMap) {
      return undefined;
    }

    if (language) {
      const registration = languageMap.get(language);
      return registration?.tool;
    }

    // Return the first available tool
    const firstRegistration = languageMap.values().next().value;
    return firstRegistration?.tool;
  }

  /**
   * Get tool handler
   */
  getHandler(toolId: string, language: Language): ToolHandler | undefined {
    const languageMap = this.tools.get(toolId);
    if (!languageMap) {
      return undefined;
    }

    const registration = languageMap.get(language);
    return registration?.handler;
  }

  /**
   * List tools with optional filter
   */
  async list(filter?: ToolFilter): Promise<Tool[]> {
    const tools: Tool[] = [];

    for (const languageMap of this.tools.values()) {
      for (const registration of languageMap.values()) {
        const tool = registration.tool;

        // Apply filters
        if (filter) {
          if (filter.category && tool.category !== filter.category) {
            continue;
          }
          if (filter.language && tool.language !== filter.language) {
            continue;
          }
          if (filter.tags && filter.tags.length > 0) {
            const toolTags = (tool.metadata?.tags as string[]) || [];
            if (!filter.tags.some((tag) => toolTags.includes(tag))) {
              continue;
            }
          }
        }

        tools.push(tool);
      }
    }

    return tools;
  }

  /**
   * Search tools by query
   */
  async search(query: string): Promise<Tool[]> {
    const lowerQuery = query.toLowerCase();
    const tools: Tool[] = [];

    for (const languageMap of this.tools.values()) {
      for (const registration of languageMap.values()) {
        const tool = registration.tool;

        // Search in name, description, and category
        if (
          tool.name.toLowerCase().includes(lowerQuery) ||
          tool.description.toLowerCase().includes(lowerQuery) ||
          tool.category.toLowerCase().includes(lowerQuery)
        ) {
          tools.push(tool);
        }
      }
    }

    return tools;
  }

  /**
   * Load built-in tools
   */
  private async loadBuiltInTools(): Promise<void> {
    // Load TypeScript tools from OpenCode
    await this.loadOpenCodeTools();

    // Load Python tools from DGM
    await this.loadDGMTools();
  }

  /**
   * Load OpenCode tools
   */
  private async loadOpenCodeTools(): Promise<void> {
    try {
      // Dynamically import OpenCode tools
      const toolModules = [
        '../../opencode/packages/opencode/src/tool/bash',
        '../../opencode/packages/opencode/src/tool/edit',
        '../../opencode/packages/opencode/src/tool/read',
        '../../opencode/packages/opencode/src/tool/write',
        '../../opencode/packages/opencode/src/tool/ls',
        '../../opencode/packages/opencode/src/tool/grep',
        '../../opencode/packages/opencode/src/tool/glob',
      ];

      for (const modulePath of toolModules) {
        try {
          const module = await import(modulePath);
          if (module.default || module.BashTool || module.EditTool) {
            // Extract tool info and register
            const tool = module.default || module.BashTool || module.EditTool;
            if (tool && tool.id) {
              await this.registerDGMOTool(tool);
            }
          }
        } catch (error) {
          console.warn(`Failed to load OpenCode tool from ${modulePath}:`, error);
        }
      }
    } catch (error) {
      console.error('Failed to load OpenCode tools:', error);
    }
  }

  /**
   * Load DGM tools
   */
  private async loadDGMTools(): Promise<void> {
    try {
      const toolModules = [
        { path: '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/dgm/tools/bash.py', name: 'bash' },
        { path: '/mnt/c/Users/jehma/Desktop/AI/DGMSTT/dgm/tools/edit.py', name: 'edit' },
      ];

      for (const { path, name } of toolModules) {
        try {
          // Load tool info from Python module
          const toolInfo = await TypeScriptPythonAdapter.loadPythonModule(path);
          if (toolInfo) {
            await this.registerDGMTool(
              {
                name,
                description: toolInfo.description || '',
                inputSchema: toolInfo.input_schema || {},
                outputSchema: toolInfo.output_schema,
                category: toolInfo.category,
                tags: toolInfo.tags,
              },
              path,
            );
          }
        } catch (error) {
          console.warn(`Failed to load DGM tool from ${path}:`, error);
        }
      }
    } catch (error) {
      console.error('Failed to load DGM tools:', error);
    }
  }

  /**
   * Infer tool category from ID
   */
  private inferCategory(toolId: string): ToolCategory {
    const categoryMap: Record<string, ToolCategory> = {
      bash: 'utility',
      edit: 'file-system',
      read: 'file-system',
      write: 'file-system',
      ls: 'file-system',
      grep: 'text-processing',
      glob: 'file-system',
      patch: 'file-system',
      multiedit: 'file-system',
    };

    return categoryMap[toolId] || 'utility';
  }

  /**
   * Convert DGMO parameters (Zod schema) to JSON Schema
   */
  private convertDGMOParametersToJsonSchema(parameters: any): any {
    // If it's already a JSON schema, return it
    if (parameters && typeof parameters === 'object' && 'type' in parameters) {
      return parameters;
    }

    // If it has a _def property, it might be a Zod schema
    if (parameters && parameters._def) {
      return TypeConverter.zodToJsonSchema(parameters);
    }

    // Default fallback
    return {
      type: 'object',
      properties: {},
      additionalProperties: true,
    };
  }

  /**
   * Get available languages for a tool
   */
  getAvailableLanguages(toolId: string): Language[] {
    const languageMap = this.tools.get(toolId);
    if (!languageMap) {
      return [];
    }

    return Array.from(languageMap.keys());
  }

  /**
   * Check if a tool supports a specific language
   */
  supportsLanguage(toolId: string, language: Language): boolean {
    const languageMap = this.tools.get(toolId);
    if (!languageMap) {
      return false;
    }

    return languageMap.has(language);
  }

  /**
   * Execute a tool with automatic language selection
   */
  async execute(
    toolId: string,
    input: any,
    context: ToolContext,
    preferredLanguage?: Language,
  ): Promise<Result<any>> {
    const languageMap = this.tools.get(toolId);
    if (!languageMap) {
      return {
        success: false,
        error: {
          code: 'TOOL_NOT_FOUND',
          message: `Tool ${toolId} not found`,
          recoverable: false,
        },
      };
    }

    // Try preferred language first
    if (preferredLanguage) {
      const registration = languageMap.get(preferredLanguage);
      if (registration) {
        return registration.handler(input, context);
      }
    }

    // Fall back to first available
    const firstRegistration = languageMap.values().next().value;
    if (firstRegistration) {
      return firstRegistration.handler(input, context);
    }

    return {
      success: false,
      error: {
        code: 'NO_HANDLER_FOUND',
        message: `No handler found for tool ${toolId}`,
        recoverable: false,
      },
    };
  }

  /**
   * Validate tool input against schema
   */
  async validateInput(toolId: string, input: any, language?: Language): Promise<Result<any>> {
    const tool = await this.get(toolId, language);
    if (!tool) {
      return {
        success: false,
        error: {
          code: 'TOOL_NOT_FOUND',
          message: `Tool ${toolId} not found`,
          recoverable: false,
        },
      };
    }

    try {
      // Validate against JSON schema
      const valid = await TypeConverter.validateAndCoerce(
        input,
        tool.inputSchema,
        language === 'python' ? 'typescript' : 'python',
      );

      return {
        success: true,
        data: valid,
      };
    } catch (error) {
      return {
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: error instanceof Error ? error.message : 'Input validation failed',
          recoverable: false,
          details: error,
        },
      };
    }
  }

  /**
   * Get tool discovery information
   */
  async discover(): Promise<{
    totalTools: number;
    byLanguage: Record<Language, number>;
    byCategory: Record<ToolCategory, number>;
    bySource: Record<string, number>;
  }> {
    const stats = {
      totalTools: 0,
      byLanguage: {} as Record<Language, number>,
      byCategory: {} as Record<ToolCategory, number>,
      bySource: {} as Record<string, number>,
    };

    for (const languageMap of this.tools.values()) {
      for (const [language, registration] of languageMap.entries()) {
        stats.totalTools++;

        // Count by language
        stats.byLanguage[language] = (stats.byLanguage[language] || 0) + 1;

        // Count by category
        const category = registration.tool.category;
        stats.byCategory[category] = (stats.byCategory[category] || 0) + 1;

        // Count by source
        const source = registration.source;
        stats.bySource[source] = (stats.bySource[source] || 0) + 1;
      }
    }

    return stats;
  }
}

// Helper function to check if a tool is a DGMO tool
function isDGMOTool(tool: any): tool is DGMOTool {
  return 'execute' in tool && typeof tool.execute === 'function';
}

// Export singleton instance
export const toolRegistry = UnifiedToolRegistry.getInstance();
